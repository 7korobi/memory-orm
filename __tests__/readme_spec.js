const { Rule, Set, Query } = require("../lib/index.min");

new Rule("todo").schema(function() {
  this.has_many("checks");
});

new Rule("check").schema(function() {
  this.belongs_to("todo");
});

Set.todo.merge([
  {
    _id: 1,
    label: "歯を磨く"
  },
  {
    _id: 2,
    label: "宿題をする"
  },
  {
    _id: 3,
    label: "お風呂はいる"
  }
]);

Set.check.merge([
  {
    _id: 1,
    todo_id: 1,
    label: "右上",
    checked: true
  },
  {
    _id: 2,
    todo_id: 1,
    label: "左上",
    checked: true
  },
  {
    _id: 3,
    todo_id: 3,
    label: "シャワー浴びる",
    checked: false
  },
  {
    _id: 4,
    todo_id: 3,
    label: "肩まで入って１０数える",
    checked: true
  }
]);


new Rule("position").schema(function() {
  this.model = class model extends this.model {
    static map_reduce (o, emit){
      for ( const p of o.position ) {
        emit("position", { count: 1, all: p, min: p, max: p})
      }
    }
  }
});

Set.position.merge([{
  "_id": "x1", 
  "position": [40,60,80,100,200]
},{
  "_id": "y1", 
  "position": [70,190,220,160,240]
},{
  "_id": "x2", 
  "position": [40,60,80,100,200]
},{
  "_id": "y2", 
  "position": [20,90,20,60,40]
}])


describe("query.todos", ()=> {
  test('labels snapshot', ()=> {
    expect(Query.todos.pluck("label")).toMatchSnapshot()
  })
  test('checked snapshot', ()=> {
    expect(Query.todos.find(3).checks.where({ checked: true }).list).toMatchSnapshot()
  })
  test('not checked snapshot', ()=> {
    expect(Query.todos.find(3).checks.where({ checked: false }).list).toMatchSnapshot()
  })
})

describe("query.positions", ()=> {
  test('x1 snapshot', ()=> {
    expect(Query.positions.where({_id: "x1"}).reduce).toMatchSnapshot()
  })
  test('checked snapshot', ()=> {
    expect(Query.positions.reduce).toMatchSnapshot()
  })
})
