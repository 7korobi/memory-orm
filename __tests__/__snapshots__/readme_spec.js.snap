// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`query.checks order asc snapshot 1`] = `
Array [
  "シャワー浴びる",
  "右上",
  "左上",
  "肩まで入って１０数える",
]
`;

exports[`query.checks order desc snapshot 1`] = `
Array [
  "肩まで入って１０数える",
  "左上",
  "右上",
  "シャワー浴びる",
]
`;

exports[`query.positions checked snapshot 1`] = `
Object {
  "hash": Object {
    "x1": model {
      "_id": "x1",
      "position": Array [
        40,
        60,
        80,
        100,
        200,
      ],
    },
    "x2": model {
      "_id": "x2",
      "position": Array [
        40,
        60,
        80,
        100,
        200,
      ],
    },
    "y1": model {
      "_id": "y1",
      "position": Array [
        70,
        190,
        220,
        160,
        240,
      ],
    },
    "y2": model {
      "_id": "y2",
      "position": Array [
        20,
        90,
        20,
        60,
        40,
      ],
    },
  },
  "list": Array [
    model {
      "_id": "x1",
      "position": Array [
        40,
        60,
        80,
        100,
        200,
      ],
    },
    model {
      "_id": "y1",
      "position": Array [
        70,
        190,
        220,
        160,
        240,
      ],
    },
    model {
      "_id": "x2",
      "position": Array [
        40,
        60,
        80,
        100,
        200,
      ],
    },
    model {
      "_id": "y2",
      "position": Array [
        20,
        90,
        20,
        60,
        40,
      ],
    },
  ],
  "position": Object {
    "all": 2070,
    "avg": 103.5,
    "count": 20,
    "density": 9.409090909090908,
    "max": 240,
    "max_is": model {
      "_id": "y1",
      "position": Array [
        70,
        190,
        220,
        160,
        240,
      ],
    },
    "min": 20,
    "min_is": model {
      "_id": "y2",
      "position": Array [
        20,
        90,
        20,
        60,
        40,
      ],
    },
    "range": 220,
  },
  "set": Array [
    "x1",
    "y1",
    "x2",
    "y2",
  ],
}
`;

exports[`query.positions x1 snapshot 1`] = `
Object {
  "hash": Object {
    "x1": model {
      "_id": "x1",
      "position": Array [
        40,
        60,
        80,
        100,
        200,
      ],
    },
  },
  "list": Array [
    model {
      "_id": "x1",
      "position": Array [
        40,
        60,
        80,
        100,
        200,
      ],
    },
  ],
  "position": Object {
    "all": 480,
    "avg": 96,
    "count": 5,
    "density": 3,
    "max": 200,
    "max_is": model {
      "_id": "x1",
      "position": Array [
        40,
        60,
        80,
        100,
        200,
      ],
    },
    "min": 40,
    "min_is": model {
      "_id": "x1",
      "position": Array [
        40,
        60,
        80,
        100,
        200,
      ],
    },
    "range": 160,
  },
  "set": Array [
    "x1",
  ],
}
`;

exports[`query.todos checked snapshot 1`] = `
Array [
  model {
    "_id": 4,
    "checked": true,
    "label": "肩まで入って１０数える",
    "todo_id": 3,
  },
]
`;

exports[`query.todos labels snapshot 1`] = `
Array [
  "歯を磨く",
  "宿題をする",
  "お風呂はいる",
]
`;

exports[`query.todos not checked snapshot 1`] = `
Array [
  model {
    "_id": 3,
    "checked": false,
    "label": "シャワー浴びる",
    "todo_id": 3,
  },
]
`;
